pkg/controllers/management/auth/globalroles/globalrole_handler_test.go

```diff
@@ -0,0 +1,834 @@
+package globalroles
+
+import (
+	"fmt"
+	"testing"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/golang/mock/gomock"
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	normanv1 "github.com/rancher/rancher/pkg/generated/norman/rbac.authorization.k8s.io/v1"
+	rbacFakes "github.com/rancher/rancher/pkg/generated/norman/rbac.authorization.k8s.io/v1/fakes"
+	"github.com/rancher/wrangler/pkg/generic/fake"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	readPodPolicyRule = rbacv1.PolicyRule{
+		Verbs:     []string{"get", "list", "watch"},
+		APIGroups: []string{""},
+		Resources: []string{"pods"},
+	}
+
+	readConfigPolicyRule = rbacv1.PolicyRule{
+		Verbs:     []string{"get", "list", "watch"},
+		APIGroups: []string{""},
+		Resources: []string{"configmaps"},
+	}
+	adminPodPolicyRule = rbacv1.PolicyRule{
+		Verbs:     []string{"*"},
+		APIGroups: []string{""},
+		Resources: []string{"pod"},
+	}
+
+	namespacedRulesGR = v3.GlobalRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "namespacedRulesGR",
+			UID:  "00000000",
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "fake-kind",
+			APIVersion: "fake-version",
+		},
+		NamespacedRules: map[string][]rbacv1.PolicyRule{
+			"namespace1": {
+				readPodPolicyRule,
+				readConfigPolicyRule,
+			},
+			"namespace2": {
+				adminPodPolicyRule,
+				readPodPolicyRule,
+			},
+		},
+	}
+	namespacedRulesOwnerRef = metav1.OwnerReference{
+		APIVersion: namespacedRulesGR.APIVersion,
+		Kind:       namespacedRulesGR.Kind,
+		Name:       namespacedRulesGR.Name,
+		UID:        namespacedRulesGR.UID,
+	}
+
+	updatedNamespacedRulesGR = v3.GlobalRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "namespacedRulesGR",
+		},
+		NamespacedRules: map[string][]rbacv1.PolicyRule{
+			"namespace1": {
+				readPodPolicyRule,
+			},
+			"namespace2": {
+				adminPodPolicyRule,
+				readPodPolicyRule,
+				readConfigPolicyRule,
+			},
+		},
+	}
+)
+
+type grTestStateChanges struct {
+	t            *testing.T
+	createdRoles map[string]*rbacv1.Role
+	deletedRoles map[string]struct{}
+}
+type grTestState struct {
+	nsCacheMock  *fake.MockNonNamespacedCacheInterface[*corev1.Namespace]
+	rListerMock  *rbacFakes.RoleListerMock
+	rClientMock  *rbacFakes.RoleInterfaceMock
+	counter      int
+	stateChanges *grTestStateChanges
+}
+
+func TestReconcileNamespacedRoles(t *testing.T) {
+	t.Parallel()
+
+	activeNamespace := &corev1.Namespace{
+		Status: corev1.NamespaceStatus{
+			Phase: corev1.NamespaceActive,
+		},
+	}
+	terminatingNamespace := &corev1.Namespace{
+		Status: corev1.NamespaceStatus{
+			Phase: corev1.NamespaceTerminating,
+		},
+	}
+
+	tests := []struct {
```

Comment:
Nit: Might be good to have a test case where we have an invalid role in a terminating namespace - we expect no new role to be created, but the existing invalid role should be deleted.

Reply:
Added